File: java-workshop/part-1-powermock/src/main/java/service/OrderService.java
package service;

import domain.Order;
import domain.PaymentResult;
import domain.PaymentStatus;
import domain.ShippingConfirmation;
import domain.StockReservation;
import repository.OrderRepository;

import java.util.List;

/**
 * Main orchestrator - creates all dependencies internally.
 * This is typical legacy code: no dependency injection, everything is instantiated inline.
 */
public class OrderService {

    public OrderResult placeOrder(Order order) {
        StockService stockService = new StockService();
        PaymentService paymentService = new PaymentService();
        ShippingService shippingService = new ShippingService();
        OrderRepository orderRepository = new OrderRepository();

        List<StockReservation> reservations = stockService.reserveStock(order.items());
        boolean hasUnavailableItems = reservations.stream().anyMatch(r -> !r.reserved());
        if (hasUnavailableItems) {
            return new OrderResult.Failure(order.reject(), "Insufficient stock");
        }

        double totalPrice = PricingService.calculateTotal(order.items());

        PaymentResult payment = paymentService.processPayment(order.id(), order.customerId(), totalPrice);
        if (payment.status() != PaymentStatus.SUCCESS) {
            stockService.releaseStock(reservations);
            return new OrderResult.Failure(order.reject(), "Payment failed");
        }

        ShippingConfirmation shipping = shippingService.createShipment(order);
        Order confirmedOrder = order.confirm(
            totalPrice,
            payment.transactionId(),
            shipping.trackingNumber()
        );

        orderRepository.save(confirmedOrder);
        return new OrderResult.Success(confirmedOrder, shipping);
    }
}

File: csharp-workshop/Part1.StaticDeps/Service/OrderService.cs
using Part1.StaticDeps.Domain;
using Part1.StaticDeps.Repository;

namespace Part1.StaticDeps.Service;

/// <summary>
/// Main orchestrator — creates all dependencies internally.
/// This is typical legacy code: no dependency injection, everything is instantiated inline.
/// </summary>
public class OrderService
{
    public OrderResult PlaceOrder(Order order)
    {
        var stockService = new StockService();
        var paymentService = new PaymentService();
        var shippingService = new ShippingService();
        var orderRepository = new OrderRepository();

        if (!IsValid(order))
        {
            return new OrderResult.Failure(order.Reject(), "Invalid order");
        }

        var reservations = stockService.ReserveStock(order.Items);
        bool hasUnavailableItems = reservations.Any(r => !r.Reserved);
        if (hasUnavailableItems)
        {
            return new OrderResult.Failure(order.Reject(), "Insufficient stock");
        }

        double totalPrice = PricingService.CalculateTotal(order.Items);

        var payment = paymentService.ProcessPayment(order.Id, order.CustomerId, totalPrice);
        if (payment.Status != PaymentStatus.Success)
        {
            stockService.ReleaseStock(reservations);
            return new OrderResult.Failure(order.Reject(), "Payment failed");
        }

        var shipping = shippingService.CreateShipment(order);
        var confirmedOrder = order.Confirm(
            totalPrice,
            payment.TransactionId!,
            shipping.TrackingNumber);

        orderRepository.Save(confirmedOrder);
        return new OrderResult.Success(confirmedOrder, shipping);
    }

    // package-private in Java → internal virtual in C# to allow NSubstitute partial mock
    // [assembly: InternalsVisibleTo("Part1.StaticDeps.Tests")] is declared in the .csproj
    internal virtual bool IsValid(Order order) =>
        order.Status == OrderStatus.Pending
        && order.Items is not null
        && order.Items.Count > 0;
}
